### PR TITLE
[WIP] Feature/speech actions

### PR DESCRIPTION
**Pull request by clement-moulin-frier**
This implements generic speech actions in the BehaviorManager. Any sentence can now be produced by the iCub by sending an rpc command to BehaviorManager. Typically:

```
yarp rpc /BehaviorManager/trigger:i
>> speech "Hello world, iCub speaking"
```

We'll use it to trigger the _"ask the human to push"_ and _"ask the human to pull"_ actions from the planner in the Y3 scenario.

### CODE REVIEW

**Comment by clement-moulin-frier** [comment]
It has been tested btw

**Comment by Tobias-Fischer** [comment]
Hmmm .. may I ask what the purpose of this is? As far as I can see, it is wrapping `iCub->say()`, why not calling it directly?

File: main/src/modules/reactiveLayer/behaviorManager/src/speech.cpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+void Speech::configure() {
+    ;
+}
+
+void Speech::run(Bottle args) {
+    yInfo() << "Speech::run";
+    // Bottle *objects = sensation.get(1).asList();
+    // Might want to include gestures like looking at partner and waving a hand
+
+    iCub->say(args.get(0).asString());
```
Comment (sockchinglow) [inline main/src/modules/reactiveLayer/behaviorManager/src/speech.cpp]: Currently, the `behaviorManager` interprets information and sends a bottle in the following manner to the behavior: `(arg1 arg2)`. I think to be compatible with that, this line should be `iCub->say(args.get(0).asList()->get(0).asString())`.

The corresponding `rpc` command to the behaviorManager would be `speech ("Hello world, iCub speaking")`.

**Comment by clement-moulin-frier** [comment]
Well, because in WR-DAC all behaviors are triggered through the BehaviorManager, so that there is a standardized way of triggering them from either the AllostaticController or from the Planner modules. In Y3 scenario those actions will be push, pull, askPull, askPush (the speech behavior is for the two last ones). 
The BM also manage safety issues so that only one behavior can be triggered at a time (mutex), preventing that a module launch e.g an ARE action whereas another one, possibly triggered from another module, is not terminated.

**Comment by Tobias-Fischer** [comment]
Hi Clement, I thought about this again, and I'm against merging this pull request. I think it is just making everyones life more complicated.
1) It adds another layer of complexity. This is easily evident from Sock's comment above, and whether this new behavior should call `iCub->say(args.get(0).asString())` or `iCub->say(args.get(0).asList()->get(0).asString())`.
2) There are already several ways in order to let the iCub say something. Either using the `iCubClient`, or issuing a RPC command to `iSpeak` (which, by the way, looks very much identical to what is proposed here). I don't see what's wrong with them.
3) You talked about the `mutex` behavior within `BehaviorManager`. This already ensures that two modules can never issue a speech command at the same time. I think the opposite is true - what if a behavior calls the new `speechBehavior`? The `mutex` will already be locked, and the `speechBehavior` will cause a deadlock. So yes, the `mutex` is needed, but not for speech.
4) Never change a running system. Everything is fine with the current system, so I wouldn't touch `iCub->say()`.

Let me know if there are issues with this.

Best, Tobi

**Comment by clement-moulin-frier** [comment]
Hi Tobi,

The thing is that we do need all behaviors involved in the planner (push, pull, askPull, askPush) to be triggered from the `BehaviorManager`, so that we can iterate through the plan and trigger the involved behaviors in a generic way. Otherwise we would have to handle each action specificity directly in the planner, it's not a good way to proceed. It's the role of the `BehaviorManager` to implement actions and to abstract them through a common interface to trigger them. I don't think implementing the actions in the planner is  a good idea, even if this action simply calls `icub->say(.)`

The same argumentation also apply regarding how behaviors are called by the `AllostaticController`.

Please tell me what you think,
Clément

**Comment by Tobias-Fischer** [comment]
Hi @clement-moulin-frier,
Let's discuss this on Wednesday during the Slack meeting if that's okay with you?

Best, Tobi

**Comment by clement-moulin-frier** [comment]
Ok but we need to merge it to be able to test the planner. I propose we do so and in the worse case it will be easy to remove (it's only a new .h and .cpp and a couple of extra lines in the `BehaviorManager` module).

But you get my point right? We need a common interface for all behaviors so that the modules that need them (allostasis and planner) can be generic. Otherwise those modules will contain behavior-specific code, which is not good.

File: main/src/modules/reactiveLayer/behaviorManager/src/speech.cpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+using namespace std;
+using namespace yarp::os;
+using namespace yarp::sig;
+
+void Speech::configure() {
+    ;
+}
+
+void Speech::run(Bottle args) {
```
Comment (pattacini) [inline main/src/modules/reactiveLayer/behaviorManager/src/speech.cpp]: Hey lads, `Bottle args` is a completely different signature from `Bottle &args` and even from `const Bottle &args`. So, have your correct choice 😉

**Comment by pattacini** [comment]
Hi @clement-moulin-frier 
I didn't follow this discussion, just want to add a quick comment.

There's no need to merge the PR to test this code. You can instead rebase this branch on top of `devel` and then use it on the robot.

**Comment by clement-moulin-frier** [comment]
Hi @Tobias-Fischer 

Sorry I missed your last comment, I was not at the lab.
We are making some tests today with the new Karma stuff in Bcn. If this work reasonably well we will then integrate it with our planner and the aforementioned speech actions. We will actually change these speech actions to more specific "Ask actions", where the robot ask something to the human while looking at him. We'll do these modifications on this branch.

So to answer your question: let's wait that we do those tests and modify the actions before merging this PR. Should be done in the next days.

Best,
Clément

File: main/src/modules/reactiveLayer/behaviorManager/src/speech.cpp (hunk 2 above)
Comment (sockchinglow) [inline main/src/modules/reactiveLayer/behaviorManager/src/speech.cpp]: This has been addressed by 21ee3a18df60bc1a49c0dd90420807b9b7fa77f7.

File: main/src/modules/reactiveLayer/behaviorManager/include/greeting.h [hunk 3]
```diff
@@ -3,15 +3,30 @@
...
-#include <map>
 
+#include <yarp/math/SVD.h>
+#include <yarp/math/Rand.h>
+#include "wrdac/clients/icubClient.h"
+#include <map>
 
 #include "behavior.h"
 
+using namespace std;
```
Comment (Tobias-Fischer) [inline main/src/modules/reactiveLayer/behaviorManager/include/greeting.h]: `using namespace` in header files is evil. Please move to the cpp file.

File: main/src/modules/reactiveLayer/behaviorManager/include/greeting.h [hunk 4]
```diff
@@ -3,15 +3,30 @@
 #include <iomanip>
 #include <yarp/os/all.h>
 #include <yarp/sig/all.h>
-#include <map>
 
+#include <yarp/math/SVD.h>
```
Comment (Tobias-Fischer) [inline main/src/modules/reactiveLayer/behaviorManager/include/greeting.h]: Is this really needed?

File: main/src/modules/reactiveLayer/behaviorManager/include/greeting.h [hunk 5]
```diff
@@ -3,15 +3,30 @@
...
+#include <yarp/math/Rand.h>
+#include "wrdac/clients/icubClient.h"
```
Comment (Tobias-Fischer) [inline main/src/modules/reactiveLayer/behaviorManager/include/greeting.h]: Please re-order the includes from general (e.g. iostream) to more and more specific (first yarp, then wysiwyd includes etc.)

File: main/src/modules/reactiveLayer/behaviorManager/include/greeting.h (hunk 3 above)
Comment (Tobias-Fischer) [inline main/src/modules/reactiveLayer/behaviorManager/include/greeting.h]: This also goes for the header files of the other behaviors ;)

File: main/src/modules/reactiveLayer/behaviorManager/src/ask.cpp [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    aux = "Can you please " + action + " the " + object + " " + direction + " me?";
+
+    iCub->lookAtPartner();
+
+    iCub->say(aux);
+    yInfo() << "iCub has requested a " + action + " from the human.";
+
+    Time::delay(2.0);
+    iCub->home();
+    Time::delay(3.5);
```
Comment (Tobias-Fischer) [inline main/src/modules/reactiveLayer/behaviorManager/src/ask.cpp]: Is there a reason for such long delays?!?

**Comment by sockchinglow** [comment]
Please note that this branch is only for the speaking actions where the iCub requests the human to manipulate objects within their `shared` and `humanOnly` spaces. Although the actions have been referenced in the planner's config file, the physical act of manipulating the objects still need to be implemented in the `BM`.

File: main/src/modules/reactiveLayer/behaviorManager/src/ask.cpp (hunk 6 above)
Comment (sockchinglow) [inline main/src/modules/reactiveLayer/behaviorManager/src/ask.cpp]: It's to ensure that the partner has time to react and shift the object before the iCub checks that the object is now in the expected region.

Comment (Tobias-Fischer) [inline main/src/modules/reactiveLayer/behaviorManager/src/ask.cpp]: Okay, I see.

File: main/src/modules/reactiveLayer/behaviorManager/include/greeting.h (hunk 5 above)
Comment (sockchinglow) [inline main/src/modules/reactiveLayer/behaviorManager/include/greeting.h]: The includes weren't necessary to begin with, so they have been removed.

File: main/src/modules/reactiveLayer/behaviorManager/src/speech.cpp (hunk 1 above)
Comment (Tobias-Fischer) [inline main/src/modules/reactiveLayer/behaviorManager/src/speech.cpp]: Only this to test before merging. All good otherwise.

**Comment by sockchinglow** [comment]
Previously, @clement-moulin-frier designed the module to work with the following arrangement:
```
yarp rpc /BehaviorManager/trigger:i
>> speech "Hello world, iCub speaking"
```

This deviated from how the `behaviorManager` sends arguments to the behaviours, and thus was modified to fit the standard handling of behaviours. Now the input should be as follows:
```
yarp rpc /BehaviorManager/trigger:i
>> speech ("Hello world, iCub speaking)"
```

**Comment by Tobias-Fischer** [comment]
Will merge after Appveyor / travis finish.

**Pull request merged**